### ISSUE TITLE
Implementation details of sparseconvolution

### ISSUE THREAD

**Issue by gopinath-r** [traveller59/spconv#231]
Hi @traveller59,

I am trying to understand your code, while doing so i am little bit confused in certain areas. If i am not wrong following are the main steps of your implementation

**1.** **Iterate once through the the input hash-table. We build the output hash table and rule
book on-the-fly by iterating over points in the output layer that receive input from a given
point in the input layer. When an output site is visited for the first time, a new entry is
created in the output hash table. Based on the spatial offset between the input and output
points, a (input index, output index) pair is added to the rule book
2. Initialize the output matrix to all zeros. For each i ∈ f, there is a parameter matrix Wi
with size m × n. For each j ∈ {1, . . . , ki}, multiply the Ri(j, 1)-th row of the input feature
matrix by Wi and add it to the Ri(j, 2)-th row of the the output feature matrix. This
can be implemented very efficiently on GPUs because it is a matrix-matrix multiply-add
operation.**

But i am not able to understand the same via your code. Can you please provide a little bit of explanation for methods Index **getIndicePairsConv** , **getValidOutPos**. Especially what is lower, upper, counter in getValidOutPos means and how their formula derived?

Basically the implementation details of first point mention above. It would be really helpfull if you can share some information or any materials that would be helpful.

Thanks in advance
